src/Query/ExecutorFactory.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace React\Dns\Query;
+
+use React\Cache\ArrayCache;
+use React\Cache\CacheInterface;
+use React\Dns\Config\Config;
+use React\Dns\Config\HostsFile;
+use React\EventLoop\Loop;
+use React\EventLoop\LoopInterface;
+
+class ExecutorFactory
+{
+
+    /**
+     * Creates a DNS executer instance for the given DNS config
+     *
+     * As of v1.7.0 it's recommended to pass a `Config` object instead of a
+     * single nameserver address. If the given config contains more than one DNS
+     * nameserver, all DNS nameservers will be used in order. The primary DNS
+     * server will always be used first before falling back to the secondary or
+     * tertiary DNS server.
+     *
+     * @param Config|string  $config DNS Config object (recommended) or single nameserver address
+     * @param ?LoopInterface $loop
+     * @return ExecutorInterface|HostsFileExecutor
+     * @throws \InvalidArgumentException for invalid DNS server address
+     * @throws \UnderflowException when given DNS Config object has an empty list of nameservers
+     */
+    public function create($config, LoopInterface $loop = null)
+    {
+        return $this->decorateHostsFileExecutor($this->createExecutor($config, $loop ?: Loop::get()));
+    }
+
+    /**
+     * Creates a cached DNS executer instance for the given DNS config and cache
+     *
+     * As of v1.7.0 it's recommended to pass a `Config` object instead of a
+     * single nameserver address. If the given config contains more than one DNS
+     * nameserver, all DNS nameservers will be used in order. The primary DNS
+     * server will always be used first before falling back to the secondary or
+     * tertiary DNS server.
+     *
+     * @param Config|string   $config DNS Config object (recommended) or single nameserver address
+     * @param ?LoopInterface  $loop
+     * @param ?CacheInterface $cache
+     * @return ExecutorInterface|HostsFileExecutor
+     * @throws \InvalidArgumentException for invalid DNS server address
+     * @throws \UnderflowException when given DNS Config object has an empty list of nameservers
+     */
+    public function createCached($config, LoopInterface $loop = null, CacheInterface $cache = null)
```

Comment:
Since we made the loop global and this is a new class it is probably favorable to put it as the last argument. What do you think @clue ?

```suggestion
    public function createCached($config, CacheInterface $cache = null, LoopInterface $loop = null)
```

Reply:
BC break?

Reply:
Agree that we want to de-emphasize the loop argument as we've started using the default loop throughout all our components. Not a BC break here given this is a new API (see above discussion).

Reply:
kk, let us change this